contracts/contracts/pools/lock_release_token_pool.tolk

```diff
@@ -0,0 +1,274 @@
+            getTokenDecimals: null,
+            encodeLocalDecimals: null,
+            calculateLocalAmount: null,
+        }
+    };
+}
+
+// TODO: handle bounced messages
+fun onInternalMessage(in: InMessage) {
+    /// Load the contract storage as MCMS and handle the message
+    var lockReleaseTokenPool = LockReleaseTokenPool.load();
+
+    if (lockReleaseTokenPool.tokenPool().onInternalMessage(in.senderAddress, in.valueCoins, in.body)) {
+        lockReleaseTokenPool.data.storeAsContractData();
+        return;
+    }
+
+    lockReleaseTokenPool.onInternalMessage(in.senderAddress, in.valueCoins, in.body);
+}
+
+
+// --- Message handlers ---
+
+
+fun LockReleaseTokenPool.onInternalMessage(mutate self, sender: address, msgValue: coins, msgBody: slice) {
+    var response: LockReleaseTokenPool_Response? = null;
+    val msg = lazy LockReleaseTokenPool_InMessage.fromSlice(msgBody);
+    match (msg) {
+        Pool_LockOrBurnInV1 => {
+            response = lockOrBurn(mutate self, sender, msgValue, msg);
+        },
+        Pool_ReleaseOrMintInV1 => {
+            response = releaseOrMint(mutate self, sender, msgValue, msg);
+        },
+        LockReleaseTokenPool_GetRebalancer => {
+            val rebalancer = self.handleGetRebalancer(sender);
+            response = LockReleaseTokenPool_GetRebalancerResponse {
+                queryId: msg.queryId,
+                rebalancer: self.getRebalancer(),
+            };
+        },
+        LockReleaseTokenPool_SetRebalancer => {
+            self.handleSetRebalancer(sender, msg.rebalancer);
+            response = LockReleaseTokenPool_SetRebalancerResponse {
+                queryId: msg.queryId,
+            };
+        },
+        // LockReleaseTokenPool_ProvideLiquidity => self.handleProvideLiquidity(sender),
+        // LockReleaseTokenPool_WithdrawLiquidity => self.handleWithdrawLiquidity(sender),
+        // LockReleaseTokenPool_TransferLiquidity => self.handleTransferLiquidity(sender),
+        else => {
+            return;
+        }
+    }
+
+    
+    createMessage({
+        bounce: false,
+        value: 0,
+        dest: sender,
+        body: response,
+    }).send(SEND_MODE_CARRY_ALL_REMAINING_MESSAGE_VALUE);
+    
+    /// Store the updated contract data
+    self.data.storeAsContractData()
+}
+
+// --- Internal functions ---
+
+// ================================================================
+// │                        Lock or Burn                          │
+// ================================================================
+/// @notice Burn the token in the pool
+/// @dev The _validateLockOrBurn check is an essential security check
+fun lockOrBurn(mutate this: LockReleaseTokenPool, sender: address, msgValue: coins, lockOrBurnIn: Pool_LockOrBurnInV1): Pool_LockOrBurnOutV1 {
+    this.tokenPool()._validateLockOrBurn(sender, lockOrBurnIn);
+
+    val token = lazy this.tokenPool().data.i_token.load();
+    emit(TOKENPOOL_LOCKEDORBURNED_TOPIC, TokenPool_LockedOrBurned {
```

Comment:
imo all of these events will need to be eventually migrated as reply - how would an user use this pool otherwise? Is it locked/burned, or is it not?